File: 0039-combination-sum/0039-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */

// either take the current candidate or the next
var combinationSum = function(candidates, target) {
    const result = []
    
    const dp = (i, total, current=[]) => {
        // base case: total === target
        // push a copy of the current array, to avoid reference issues
        if (total === target) {
            result.push(current.slice())
            return;
        }
        
        // base case: if total exceeds target or run out of candidates, there is no need to recursive
        if (i >= candidates.length || total > target) {
            return;
        }
        
        // recursive: take candidates[i] FIRST
        current.push(candidates[i])
        dp(i, total + candidates[i], current)
        
        // recursive: skip candidates[i] and take the next candidate
        current.pop()
        dp(i + 1, total, current)
    }
    
    dp(0, 0)
    return result;
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        results = []
        
        def backtrack(current_combination, current_index, remaining):
            # base case: reached target, append a copy of current_combination to result
            if remaining == 0:
                results.append(current_combination.copy())
                return
            # base case: ran out of candidates or overshot the target, return
            if remaining < 0 or current_index >= len(candidates):
                return
            
            current_candidate = candidates[current_index]
            
            # decision: take the current candidate
            current_combination.append(current_candidate)
            backtrack(current_combination, current_index, remaining - current_candidate)
            
            # decision: skip current candidate (pop it out of current_combination)
            current_combination.pop()
            backtrack(current_combination, current_index + 1, remaining)
            
        backtrack([], 0, target)
        return results